Fix/add filesystem watcher event to queue

amanPras opened the pull request:
Bring back https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/pull/2443

amanPras committed:
fix: add file watcher events to queue

This reverts commit 6ca680bf6fbc093f95e88b254955241d27fa86f3.

amanPras committed:
Reapply "chore: update vscode test package"

This reverts commit 64271f80c5bc20dfca352125f345824c4481f3e1.

amanPras committed:
Reapply "chore: add debug logger"

This reverts commit ef2c2bdaf6eb296529119a66f85e0b7a8f637346.

amanPras committed:
test: fix integration tests

Signed-off-by: ap891843 <aman.prashant@broadcom.com>

ap891843 commented on clients/cobol-lsp-vscode-extension/src/test/suite/lsp.spec.test.ts:
```diff
@@ -151,7 +155,7 @@ suite("Integration Test Suite", function () {
       range(pos(58, 21), pos(58, 28)),
     );
-    assert.ok(diagnostics[0].message.includes("Syntax error on 'XCTL123'"));
+    assert.ok(diagnostics[0].message.includes("Extraneous input 'XCTL123'"));
```
I believe, this is some recent change

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/test/suite/lsp.spec.test.ts:
```diff
@@ -401,16 +399,15 @@ suite("Integration Test Suite", function () {
+      () => vscode.languages.getDiagnostics(editor.document.uri).length > 0,
     );
     let diagnostics = vscode.languages.getDiagnostics(editor.document.uri);
     assert.strictEqual(
```
Can you be 100% sure that the message you are waiting for will be the first entry in the array? Should the code search all the diags and try to find it?

amanPras replied:
updated

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/test/suite/testHelper.ts:
```diff
@@ -78,31 +78,30 @@ export async function showDocument(workspace_file: string) {
-    if (vscode.window.activeTextEditor === undefined) {
-      return true;
-    }
+  let doc = vscode.window.activeTextEditor;
```
```suggestion
  let editor = vscode.window.activeTextEditor;
```
or maybe just 
```
while(vscode.window.activeTextEditor?.document)
```

amanPras replied:
updated

slavek-kucera commented on server/common/src/main/java/org/eclipse/lsp/cobol/common/benchmark/BenchmarkServiceImpl.java:
```diff
@@ -24,12 +24,13 @@
   private final static Logger LOG = LoggerFactory.getLogger(BenchmarkServiceImpl.class);
   private static final String PERFORMANCE_LOG_PATH = "performance.log.path";
-  private final List<BenchmarkSession> benchmarkSessions = new ArrayList<>();
+  private final List<BenchmarkSession> benchmarkSessions = new CopyOnWriteArrayList<>();
```
What code accesses this varible concurrently?

amanPras replied:
https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/blob/295158b35558a5aa30c26bbf4d9a91e141bc06a7/server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java#L165

This service is invoked by each analysis thread. Multiple thread could try to logtime

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java:
```diff
@@ -114,19 +114,36 @@ public synchronized CompletableFuture<CobolDocumentModel> scheduleAnalysis(Strin
-    Executor analysisExecutor = getExecutor(uri);
-    CompletableFuture<CobolDocumentModel> value = CompletableFuture.supplyAsync(() -> {
+    ExecutorService analysisExecutor = getExecutor(uri);
+    CobolDocumentModel documentModel = documentModelService.get(uri);
+    documentModel.updateLastAnalysisNotCancelled();
+    FutureTask<CobolDocumentModel> futureTask = new FutureTask<>(scheduleAnalysis(uri, documentModel, text, currentRevision, open, force, eventSource, id));
+    analysisResults.put(id, futureTask);
+    analysisExecutor.submit(futureTask);
+    if (prevId != null && !force) {
+      Optional.ofNullable(analysisResults.get(makeId(uri, prevId))).ifPresent(cf -> cf.cancel(true));
+    }
```
Shouldn't this be in a similar loop that can be found in `reanalyseOpenedPrograms`?

amanPras replied:
I am under impression that this method is responsible to scheduling an analysis and the caller of this method should try to cancel previous analysis before calling this method.
But I think, `reanalyseCopybooksAssociatedPrograms` should be updated

slavek-kucera replied:
Fine with me, but be also aware of the other direct calls to the `scheduleAnalysis` function - from didOpen, didChange handlers and AnalysisHandler.getDependencies.

slavek-kucera replied:
to be handled in #2467.

amanPras committed:
fix: update as per PR comments

Signed-off-by: ap891843 <aman.prashant@broadcom.com>

amanPras committed:
fix: update as per PR comments

Signed-off-by: ap891843 <aman.prashant@broadcom.com>

amanPras committed:
Revert "fix: update as per PR comments"

This reverts commit 5fa3c988d6f70f82116e6403b9475c8d1e890f51.

amanPras committed:
fix: cancel previous analysis before triggering a new analysis

Signed-off-by: ap891843 <aman.prashant@broadcom.com>

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java:
```diff
@@ -114,47 +116,71 @@ public synchronized CompletableFuture<CobolDocumentModel> scheduleAnalysis(Strin
-      notifyAllListeners(AnalysisState.STARTED, documentModelService.get(uri), eventSource);
+      LOG.debug("[scheduleAnalysis] waiting for previous analysis of {} to finish", uri);
       try {
+        if (documentModel.getLastAnalysisResult() != null) {
+          waitForPreviousAnalysisToBeCompleted(ImmutableList.of(uri));
+        }
```
Should the waiting be unconditional? The rule is that we don't want to cancel the first parsing, however we still don't want run two parsings side by side, right?

amanPras replied:
Yes, totally missed this scenario.

Updated in the latest commit

amanPras committed:
fix: cancel previous analysis before triggering a new analysis

Signed-off-by: ap891843 <aman.prashant@broadcom.com>

amanPras commented:
waiting for manual test

amanPras committed:
chore: add debug logger

Signed-off-by: ap891843 <aman.prashant@broadcom.com>

amanPras committed:
fix: few small fixes

following small fixes:
1. Event source are properly propagated for file change events
2. hover should work for copybooks

Signed-off-by: ap891843 <aman.prashant@broadcom.com>

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java:
```diff
@@ -154,21 +154,22 @@ private Callable<CobolDocumentModel> scheduleAnalysis(String uri, CobolDocumentM
       try {
         waitForPreviousAnalysisToBeCompleted(ImmutableList.of(uri));
-        documentModel.markAnalysisInProgress();
+        documentModelService.get(uri).markAnalysisInProgress();
```
Given the multithreading nature of this code, anytime you call the `documentModelService.get(uri)` you might get a different result, right?

amanPras replied:
Yes, I will revert this

amanPras replied:
updated

amanPras committed:
feat: simplify synchronization among analysis thread

Signed-off-by: ap891843 <aman.prashant@broadcom.com>
